Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +90,769 @@ local macro "pderiv_simp" : tactic =>
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 ≠ Q y * P z ^ 3) : P y ≠ V.negY P := by
+  have hy' : P y * Q z ^ 3 - V.negY Q * P z ^ 3 = 0 :=
+    (mul_eq_zero.mp <| Y_sub_Y_mul_Y_sub_negY hP hQ hx).resolve_left <| sub_ne_zero_of_ne hy
+  contrapose! hy
+  linear_combination (norm := ring1) Y_sub_Y_add_Y_sub_negY P Q hx + Q z ^ 3 * hy - hy'
 
-end Equation
+lemma Y_ne_negY_of_Y_ne' [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : V.Equation P)
+    (hQ : V.Equation Q) (hx : P x * Q z ^ 2 = Q x * P z ^ 2)
+    (hy : P y * Q z ^ 3 ≠ V.negY Q * P z ^ 3) : P y ≠ V.negY P := by
+  have hy' : P y * Q z ^ 3 - Q y * P z ^ 3 = 0 :=
+    (mul_eq_zero.mp <| Y_sub_Y_mul_Y_sub_negY hP hQ hx).resolve_right <| sub_ne_zero_of_ne hy
+  contrapose! hy
+  linear_combination (norm := ring1) Y_sub_Y_add_Y_sub_negY P Q hx + Q z ^ 3 * hy - hy'
+
+lemma Y_eq_negY_of_Y_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hQz : Q z ≠ 0)
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3)
+    (hy' : P y * Q z ^ 3 = V.negY Q * P z ^ 3) : P y = V.negY P :=
+  mul_left_injective₀ (pow_ne_zero 3 hQz) <| by
+    linear_combination (norm := ring1) -Y_sub_Y_add_Y_sub_negY P Q hx + hy + hy'
+
+lemma nonsingular_iff_of_Y_eq_negY {P : Fin 3 → F} (hPy : P y = W.negY P) (hPz : P z ≠ 0) :
+    W.Nonsingular P ↔ W.Equation P ∧ eval P W.polynomialX ≠ 0 := by
+  rw [nonsingular_iff_of_Z_ne_zero hPz, ← Y_ne_negY_iff, hPy, ne_self_iff_false, or_false]
+
+end Negation
+
+section Doubling
+
+/-! ### Doubling formulae -/
+
+variable (V) in
+/-- The $Z$-coordinate of the doubling of a point representative. -/
+def dblZ (P : Fin 3 → R) : R :=
+  P z * (P y - V.negY P)
+
+lemma dblZ_smul (P : Fin 3 → R) (u : Rˣ) : V.dblZ (u • P) = u ^ 4 * V.dblZ P := by
+  simp only [dblZ, negY_smul, smul_fin3_ext]
+  ring1
+
+lemma dblZ_of_Z_eq_zero {P : Fin 3 → R} (hPz : P z = 0) : V.dblZ P = 0 := by
+  rw [dblZ, hPz, zero_mul]
+
+lemma dblZ_of_Y_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hQz : Q z ≠ 0)
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3)
+    (hy' : P y * Q z ^ 3 = V.negY Q * P z ^ 3) : V.dblZ P = 0 := by
+  rw [dblZ, Y_eq_negY_of_Y_eq hQz hx hy hy', sub_self, mul_zero]
+
+lemma dblZ_ne_zero_of_Y_ne [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : V.Equation P)
+    (hQ : V.Equation Q) (hPz : P z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2)
+    (hy : P y * Q z ^ 3 ≠ Q y * P z ^ 3) : V.dblZ P ≠ 0 :=
+  mul_ne_zero hPz <| sub_ne_zero_of_ne <| Y_ne_negY_of_Y_ne hP hQ hx hy
+
+lemma dblZ_ne_zero_of_Y_ne' [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : V.Equation P)
+    (hQ : V.Equation Q) (hPz : P z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2)
+    (hy : P y * Q z ^ 3 ≠ V.negY Q * P z ^ 3) : V.dblZ P ≠ 0 :=
+  mul_ne_zero hPz <| sub_ne_zero_of_ne <| Y_ne_negY_of_Y_ne' hP hQ hx hy
+
+private lemma toAffine_slope_of_eq {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q)
```

Comment:
Why is this private? You're guaranteeing that nobody else will ever need this?

Reply:
Ditto.

Reply:
(for anyone else reading this: the answer is "there's a better lemma later")